File: tests/e2e/assert_firefox_container.py
#!/usr/bin/env python3
"""Assert one browser-produced Firefox container cookie through Python."""

from __future__ import annotations

import argparse
import sys

import rookie_cookies


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("database")
    parser.add_argument("user_context_id", type=int)
    parser.add_argument(
        "origin_attributes",
        help="the verbatim originAttributes suffix Firefox wrote for this row",
    )
    args = parser.parse_args()
    snapshot = rookie_cookies.from_path(args.database)
    context = {
        "url": "https://container.rookie.test/",
        "top_level_site": "https://container.rookie.test/",
        "user_context_id": args.user_context_id,
    }
    if snapshot.header(context) != "rookie_container=container-1":
        raise RuntimeError("Python did not select the exact Firefox container cookie")
    if snapshot.header({**context, "user_context_id": args.user_context_id + 1}) != "":
        raise RuntimeError("Python merged a different Firefox container")

    view = snapshot.send_view(context)
    identities = [
        (record["cookie"]["name"], record["cookie"]["value"])
        for record in view["cookies"]
    ]
    if identities != [("rookie_container", "container-1")]:
        raise RuntimeError(f"Python send_view selected {identities!r}")
    if view["header"] != "rookie_container=container-1":
        raise RuntimeError(f"Python send_view header was {view['header']!r}")
    # The raw suffix is never a bypass: the typed container selector still has
    # to match, so naming the stored suffix must land on the same single record
    # rather than widening the set.
    round_trip = snapshot.send_view(
        {**context, "origin_attributes": args.origin_attributes}
    )
    if round_trip["cookies"] != view["cookies"]:
        raise RuntimeError(
            f"the raw origin-attribute selector {args.origin_attributes!r} changed "
            f"the selected set: {round_trip['cookies']!r}"
        )

    try:
        snapshot.header(
            {
                "url": context["url"],
                "top_level_site": context["top_level_site"],
            }
        )
    except Exception as error:
        if getattr(error, "code", None) != "incomplete_send_context" and (
            "incomplete_send_context" not in str(error)
        ):
            raise
        required = list(getattr(error, "required", None) or [])
        if required != ["user_context_id"]:
            raise RuntimeError(
                f"Python named the wrong required selectors: {required!r}"
            ) from error
    else:
        raise RuntimeError("Python accepted a missing Firefox container selector")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as error:
        print(f"Firefox container Python assertion failed: {error}", file=sys.stderr)
        raise

File: tests/e2e/assert_firefox_container.mjs
// Assert one browser-produced Firefox container cookie through Node.

import process from "node:process";

import * as rookieCookies from "../../bindings/node/index.js";

const [database, idArg, originAttributes] = process.argv.slice(2);
const userContextId = Number(idArg);
if (
  !database ||
  !Number.isSafeInteger(userContextId) ||
  userContextId <= 0 ||
  !originAttributes
) {
  console.error(
    "usage: node assert_firefox_container.mjs <database> <user-context-id> <origin-attributes>",
  );
  process.exit(2);
}

const snapshot = await rookieCookies.fromPath({ path: database });
const context = {
  url: "https://container.rookie.test/",
  topLevelSite: "https://container.rookie.test/",
  userContextId,
};
if (snapshot.header(context) !== "rookie_container=container-1") {
  throw new Error("Node did not select the exact Firefox container cookie");
}
if (snapshot.header({ ...context, userContextId: userContextId + 1 }) !== "") {
  throw new Error("Node merged a different Firefox container");
}

const view = snapshot.sendView(context);
const identities = view.cookies.map(({ cookie }) => [cookie.name, cookie.value]);
if (
  JSON.stringify(identities) !==
  JSON.stringify([["rookie_container", "container-1"]])
) {
  throw new Error(`Node sendView selected ${JSON.stringify(identities)}`);
}
if (view.header !== "rookie_container=container-1") {
  throw new Error(`Node sendView header was ${view.header}`);
}
// The raw suffix is never a bypass: the typed container selector still has to
// match, so naming the stored suffix must land on the same single record
// rather than widening the set.
const roundTrip = snapshot.sendView({ ...context, originAttributes });
if (JSON.stringify(roundTrip.cookies) !== JSON.stringify(view.cookies)) {
  throw new Error(
    `the raw origin-attribute selector ${originAttributes} changed the selected set: ${JSON.stringify(roundTrip.cookies)}`,
  );
}

try {
  snapshot.header({
    url: context.url,
    topLevelSite: context.topLevelSite,
  });
  throw new Error("Node accepted a missing Firefox container selector");
} catch (error) {
  if (
    error.rookieCode !== "incomplete_send_context" &&
    !String(error).includes("incomplete_send_context")
  ) {
    throw error;
  }
  if (JSON.stringify(error.required ?? []) !== JSON.stringify(["user_context_id"])) {
    throw new Error(
      `Node named the wrong required selectors: ${JSON.stringify(error.required)}`,
    );
  }
}
